### ISSUE TITLE
Serious performance issues with PointCloud creation

### ISSUE THREAD

**Issue by fran6co** [PointCloudLibrary/pcl#1306]
Hi, while profiling some code I found out that the creation of the PointCloud structure takes too much time. Digging in I found out that the problem is the std::vector for the points. For some reason std::vector and Eigen don't work that well together.

I did a simple code to show this issue:

```
PointCloud: 0.0726809s
Array: 0.0503396s
```

A simple array is 30% faster than the PointCloud. Would it be crazy to remove std::vector from the PointCloud structure?

```
#include <iostream>
#include <chrono>

#include <pcl/common/common.h>

const int width = 1920, height = 1080;
const int cycles = 1000;

void pointCloud()
{
    auto total = std::chrono::duration<double>::zero();
    for (int i = 0; i < cycles; i++) {
        auto start = std::chrono::high_resolution_clock::now();

        {
            pcl::PointCloud<pcl::PointXYZRGBA>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZRGBA>(width, height));

            auto& points = cloud->points;
            for(int y=0, i=0;y<height;y++){
                for(int x=0;x<width;x++,i++) {
                    pcl::PointXYZRGBA& point = points[i];

                    point.x = NAN;
                    point.y = NAN;
                    point.z = NAN;
                    point.r = 0;
                    point.g = 0;
                    point.b = 0;
                    point.a = 0;
                }
            }
        }

        total += std::chrono::high_resolution_clock::now() - start;
    }
    std::cout << "PointCloud: " << total.count() / cycles << "s" << std::endl;
}

void array()
{
    auto total = std::chrono::duration<double>::zero();
    for (int i = 0; i < cycles; i++) {
        auto start = std::chrono::high_resolution_clock::now();

        {
            auto points = new pcl::PointXYZRGBA[width * height];

            for(int y=0, i=0;y<height;y++){
                for(int x=0;x<width;x++,i++) {
                    pcl::PointXYZRGBA& point = points[i];

                    point.x = NAN;
                    point.y = NAN;
                    point.z = NAN;
                    point.r = 0;
                    point.g = 0;
                    point.b = 0;
                    point.a = 0;
                }
            }

            delete[] points;
        }

        total += std::chrono::high_resolution_clock::now() - start;
    }
    std::cout << "Array: " << total.count() / cycles << "s" << std::endl;
}

int main(int argc, const char * argv[]) {

    pointCloud();
    array();

    return 0;
}
```

**Comment by taketwo** [comment]
I investigated this issue some time ago. The difference is that `std::vector` initializes data, whereas `new` does not. Unfortunately there is no way to tell vector not to do so.

Getting rid of it sounds crazy indeed. At the very least, `points` vector is a public member of `PointCloud`. If we change this, we will break all sorts of compatibilities. So not until PCL2.0 (if it ever gonna happen :smile:)

**Comment by fran6co** [comment]
Is it possible to create a new data structure that inherits from std::vector and change it so it doesn't initialize all the points?

**Comment by fran6co** [comment]
Bah, it wouldn't work... I think the PointCloud class interface needs to change so it doesn't expose the inner structure.

**Comment by fran6co** [comment]
I'm thinking this backward compatible solution:
 * Create setters & getters for all the public members in the PointCloud
 * Change all the code to use them
 * Change the interface to hide the std::vector implementation (iterators and pointers like Opencv?)
 * Create an abstract class that PointCloud inherits from and make all the code use that one instead
 * Create a PointCloud that uses an array instead of vector without public members
 * For 2.0 remove the std::vector PointCloud

**Comment by taketwo** [comment]
Second point will be problematic. Sure, we can change the code in PCL, but there are gazillions of lines of user code that use `points`. We can not force people to change all of it just to make their apps compatible with next minor release of PCL.

**Comment by fran6co** [comment]
That's why that change would be in 2.0, for now it would have the new interface and the old one. If you use code with the new interface you can use array implementation instead.

**Comment by taketwo** [comment]
Ok. But then if all the PCL suddenly starts to interact with point clouds through a base class pointer, then we will experience a performance hit due to virtual functions. (It's a standard practice to use `pointcloud->at(index)` to access elements one after another. Just imagine each of these will be a virtual call).

**Comment by jspricke** [comment]
Wouldn't it be enough to use reserve instead of resize? I'm not near a computer atm, so I can't try.

**Comment by fran6co** [comment]
Not sure if that performance hit would as bad as the one we are having now. I mean, this is a big performance hit for real time apps that create pointclouds with every frame of the camera. I don't think it would need to use the ```at``` operator every time if the code iterates over all the points, even for random access a pointer to the points would be enough.

**Comment by fran6co** [comment]
I did try the reserve and resize, no difference overall =/

**Comment by taketwo** [comment]
>  don't think it would need to use the at operator every time if the code iterates over all the points

Yes, but users who have such code will suffer. Also code in PCL that follows this pattern will need to be updated.

Another thing is that our interfaces in PCL are not always working with pointers to point clouds. In many cases we take point clouds by reference. This will make it problematic to adopt base interface class.

**Comment by fran6co** [comment]
If my memory serves me well, I think reference works fine with base interfaces the problem is with by value.

The at operator could be implemented as an inline method in the Base class that just does:

```
inline PointType at(int index)
{
    PointType* pointer = ptr<PointType>();
    return pointer[index];
}

// For std::vector
PointType* ptr()
{
   return &points[0];
}
```

**Comment by taketwo** [comment]
> If my memory serves me well, I think reference works fine with base interfaces the problem is with by value.

Sorry, you are correct.

> The at operator could be implemented as an inline method in the Base class that just does:

I don't see how this helps, the `ptr()` will need to be virtual. Whatever indirection you invent, there will be a virtual call somewhere.

There is another potential issue with this particular code that you gave. We use vectors with aligned allocators. I am not 100% sure, but suspect that depending on the point type size we will get different results accessing elements through `operator[]` vs. raw pointer.

**Comment by fran6co** [comment]
You are right, in any case I'm not sure if the virtual call penalty would be such a big issue, I think the compiler can optimize it away by inlining it if it can infer by the context that the class doesn't change (I think it would be the case).

**Comment by fran6co** [comment]
@jspricke Just updated the code and times with reserve.

**Comment by soyersoyer** [comment]
Why don't you use the vector's fill constructor? Or the vector::resize second parameter?

**Comment by fran6co** [comment]
@soyersoyer Because I don't want to fill it with the same value, the normal use case is that you have to create the point cloud from some other data structures.

**Comment by soyersoyer** [comment]
Okey. There is a range constructor and range assign too.

**Comment by fran6co** [comment]
Still, you get arbitrary data from a RGB array and you have to put it inside the vector. Not sure how it would help with a range constructor/assign. Please provide an example so I can profile it.

**Comment by fran6co** [comment]
@soyersoyer, just updated the code with a real use case. That's the code I'm using for creating a cloud out of a Kinect raw data sources.

**Comment by soyersoyer** [comment]
Sorry, I don't have pcl at this time on this computer. But this is the idea:
```
struct XYZRGB {
        float x,y,z,r,g,b;
        XYZRGB(float r,float g,float b):x(),y(),z(),r(r),g(g),b(b) {}
};
struct RGB {
        uint8_t r,g,b;
        operator XYZRGB() {
            return XYZRGB(r,g,b);
        }
};

int main() {
    uint8_t* raw_data = new uint8_t[3000];
    RGB* a = reinterpret_cast<RGB*>(raw_data);
    std::vector<XYZRGB> x(a,a+1000);
}
```
You can cast your input data to your layout compatible 'smarter' data type, which knows how to convert itself to the desired type.
If you have two input data source, you can use a simple second loop to copy that data.

**Comment by fran6co** [comment]
Ok, I'll try to create some code with that and see how it goes but I don't think it would be much better.

**Comment by fran6co** [comment]
Not sure if you wanted to do something like this, but the numbers are not that good. Not sure if it's only the initialization of the points, I think the problem is that std::vector calls the aligened_allocator of Eigen.

```
PointCloud Resize: 0.0901354s
PointCloud Reserve: 0.12211s
PointCloud Range: 0.109507s
Array: 0.0564342s
```

```c++
void pointCloudRange(uint8_t* _rgb, uint16_t* depthMap)
{
    auto total = std::chrono::duration<double>::zero();
    for (int c = 0; c < cycles; c++) {
        auto start = std::chrono::high_resolution_clock::now();

        {
            RGB* rgb = reinterpret_cast<RGB*>(_rgb);
            std::vector<pcl::PointXYZRGBA> points (rgb, rgb+width*height);

            for(int y=0, i=0;y<height;y++){
                for(int x=0;x<width;x++,i++) {
                    pcl::PointXYZRGBA& point = points[i];

                    point.x = NAN;
                    point.y = NAN;
                    point.z = NAN;

                    point.a = 255;

                    uint16_t depth = depthMap[i];
                    if (depth != 0) {
                        double zFactor = zFactors[depth];

                        point.x = static_cast<float>((x - cx) * zFactor * 0.001);
                        point.y = static_cast<float>((y - cy) * zFactor * 0.001);
                        point.z = static_cast<float>(depth * 0.001);

                        point.a = 0;
                    }
                }
            }
        }

        total += std::chrono::high_resolution_clock::now() - start;
    }
    std::cout << "PointCloud Range: " << total.count() / cycles << "s" << std::endl;
}
```

**Comment by soyersoyer** [comment]
Did you enable the optimalizations? (-O2, -O3)

**Comment by fran6co** [comment]
Yes, I'm using O3

**Comment by ahundt** [comment]
I suggest taking a look at how optimized utilities in other libraries work, and allow the implementation details of allocation to be abstracted away. In particular [boost::container](https://github.com/boostorg/container), and [boost::geometry](https://github.com/boostorg/geometry), they illustrate how to properly implement this sort of thing using a policy template so all the crucial details can be modified for performance specific applications. CRTP is greatly preferable to base classes if performance is critical.

**Comment by fran6co** [comment]
Or maybe doing something like opencv and do it in a custom way. In any case, for any solution I think the interface of the PointClouds should be independent of the implementation, right now it's exposing std::vector everywhere. When I get some extra time I'll finish my PR (help is welcomed)

**Comment by morrisf** [comment]
I would like to point out that using by using gcc4.9 with `-march=native -m64`, the compiler will optimise the resize function to match the speed of the array function:
```
PointCloud Resize: 0.0235383s
PointCloud Reserve: 0.0262214s
Array: 0.023975s
```

But by taking advantage of the initialisation function of std::vector, it is possible to improve even further, by initialising the pointcloud with points where x = y = z = NaN, and a = 255. As such, taking the constants out of the for-loop.
```
void pointCloudInit(uint8_t* rgb, uint16_t* depthMap)
{
    auto total = std::chrono::duration<double>::zero();
    for (int c = 0; c < cycles; c++) {
        auto start = std::chrono::high_resolution_clock::now();

        {
            pcl::PointXYZRGBA point_default;
            point_default.x = point_default.y = point_default.z = NAN;
            point_default.a = 255;
            std::vector<pcl::PointXYZRGBA> points(width*height, point_default);

            for(int y=0, i=0;y<height;y++){
                for(int x=0;x<width;x++,i++) {
                    pcl::PointXYZRGBA &point = points[i];

                    point.r = rgb[3*i];
                    point.g = rgb[3*i+1];
                    point.b = rgb[3*i+2];

                    uint16_t depth = depthMap[i];
                    if (depth != 0) {
                    	double zFactor = zFactors[depth];

                        point.x = static_cast<float>((x - cx) * zFactor * 0.001);
                        point.y = static_cast<float>((y - cy) * zFactor * 0.001);
                        point.z = static_cast<float>(depth * 0.001);

                        point.a = 0;
                    }
                }
            }
        }

        total += std::chrono::high_resolution_clock::now() - start;
    }
    std::cout << "PointCloud Init: " << total.count() / cycles << "s" << std::endl;
}
```
could improve over the array function with roughly 10%
`PointCloud Init: 0.0217414s`

**Comment by fran6co** [comment]
That's great! I'll check with a newer version of clang too.
